File: go/builtins/funcLength.go
package builtins

import (
	jpl "github.com/jplorg/jpl/go/v2/jpl"
	library "github.com/jplorg/jpl/go/v2/library"
)

var funcLength = enclose(func(runtime jpl.JPLRuntime, signal jpl.JPLRuntimeSignal, next jpl.JPLPiper, input any, args ...any) ([]any, error) {
	value, err := library.UnwrapValue(input)
	if err != nil {
		return nil, err
	}
	t, err := library.Type(value)
	if err != nil {
		return nil, err
	}

	switch t {
	case jpl.JPLT_NULL:
		return next.Pipe(0.)

	case jpl.JPLT_ARRAY:
		return next.Pipe(float64(len(value.([]any))))

	case jpl.JPLT_STRING:
		return next.Pipe(float64(len([]rune(value.(string)))))

	case jpl.JPLT_OBJECT:
		return next.Pipe(float64(len(value.(map[string]any))))

	default:
	}

	return nil, library.ThrowAny(library.NewTypeError("%s (%*<100v) has no length", string(t), value))
})

File: js/lib/builtins/funcLength.ts
import type JPLRuntimeSignal from '@/library/runtimeSignal';
import type JPLRuntime from '@/runtime';
import { JPLTypeError } from '../library';

export default function builtin(
  runtime: JPLRuntime,
  _signal: JPLRuntimeSignal,
  next: (output: unknown) => Promise<unknown[]> | unknown[],
  input: unknown,
) {
  const value = runtime.unwrapValue(input);
  const t = runtime.type(value);

  switch (t) {
    case 'null':
      return next(0);

    case 'array':
      return next((value as unknown[]).length);

    case 'string':
      return next([...(value as string)].length);

    case 'object':
      return next(Object.keys(value as Record<string, unknown>).length);

    default:
  }

  throw new JPLTypeError('%s (%*<100v) has no length', t, value);
}
